tests/rkt_api_service_test.go

```diff
@@ -0,0 +1,297 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"syscall"
+	"testing"
+
+	"github.com/coreos/rkt/Godeps/_workspace/src/github.com/steveeJ/gexpect"
+	"github.com/coreos/rkt/Godeps/_workspace/src/golang.org/x/net/context"
+	"github.com/coreos/rkt/Godeps/_workspace/src/google.golang.org/grpc"
+	"github.com/coreos/rkt/api/v1alpha"
+	"github.com/coreos/rkt/common"
+)
+
+func startAPIService(t *testing.T, ctx *rktRunCtx) *gexpect.ExpectSubprocess {
+	noGid := false
+	gid, err := common.LookupGid(common.RktGroup)
+	if err != nil {
+		t.Logf("no %q group, will run api service with root, ONLY DO THIS FOR TESTING!", common.RktGroup)
+		noGid = true
+	}
+
+	installCmd := fmt.Sprintf("%s install", ctx.cmd())
+	t.Logf("Running rkt install")
+	runRktAndCheckOutput(t, installCmd, "rkt directory structure successfully created", false)
+
+	t.Logf("Running rkt api service")
+	apisvcCmd := fmt.Sprintf("%s api-service", ctx.cmd())
+	if noGid {
+		return startRktAndCheckOutput(t, apisvcCmd, "API service running", false)
+	}
+	return startRktAsGidAndCheckOutput(t, apisvcCmd, "API service running", false, gid)
+}
+
+func stopAPIService(t *testing.T, svc *gexpect.ExpectSubprocess) {
+	if err := svc.Cmd.Process.Signal(syscall.SIGINT); err != nil {
+		t.Fatalf("Failed to stop the api service")
+	}
+	waitOrFail(t, svc, true)
+}
+
+func checkPodState(t *testing.T, rawState string, apiState v1alpha.PodState) {
+	switch rawState {
+	case "embryo":
+		if apiState == v1alpha.PodState_POD_STATE_EMBRYO {
+			return
+		}
+	case "preparing":
+		if apiState == v1alpha.PodState_POD_STATE_PREPARING {
+			return
+		}
+	case "aborted prepare":
+		if apiState == v1alpha.PodState_POD_STATE_ABORTED_PREPARE {
+			return
+		}
+	case "running":
+		if apiState == v1alpha.PodState_POD_STATE_RUNNING {
+			return
+		}
+	case "deleting":
+		if apiState == v1alpha.PodState_POD_STATE_DELETING {
+			return
+		}
+	case "exited":
+		if apiState == v1alpha.PodState_POD_STATE_EXITED {
+			return
+		}
+	case "garbage":
+		if apiState == v1alpha.PodState_POD_STATE_GARBAGE {
+			return
+		}
+		t.Fatalf("Unexpected state: %v", rawState)
+	}
+	t.Errorf("Pod state returned by api-service (%q) is not equivalent to the state returned by 'rkt status' (%q)", apiState, rawState)
+}
+
+func checkPodApps(t *testing.T, rawApps []*appInfo, apiApps []*v1alpha.App, checkAppState bool) {
+	if len(rawApps) != len(apiApps) {
+		t.Errorf("Expected %d apps, saw %d apps returned by api service %v", len(rawApps), len(apiApps), apiApps)
+	}
+
+	for _, m := range rawApps {
+		found := false
+		for _, n := range apiApps {
```

Comment:
How about iterating `apiApps` only once, to produce a `map[string]*v1alpha.App`?